be/src/exec/pipeline/query_context.cpp

```diff
@@ -28,43 +28,121 @@ void QueryContext::cancel(const Status& status) {
     _fragment_mgr->cancel(status);
 }
 
-QueryContextManager::QueryContextManager() = default;
+static constexpr size_t QUERY_CONTEXT_MAP_SLOT_NUM = 64;
+static constexpr size_t QUERY_CONTEXT_MAP_SLOT_NUM_MASK = (1 << 6) - 1;
+QueryContextManager::QueryContextManager()
+        : _mutexes(QUERY_CONTEXT_MAP_SLOT_NUM),
+          _context_maps(QUERY_CONTEXT_MAP_SLOT_NUM),
+          _second_chance_maps(QUERY_CONTEXT_MAP_SLOT_NUM) {}
+
+#ifdef BE_TEST
+QueryContextManager::QueryContextManager(int) : QueryContextManager() {}
+#endif
+
 QueryContextManager::~QueryContextManager() = default;
 QueryContext* QueryContextManager::get_or_register(const TUniqueId& query_id) {
-    std::lock_guard lock(_lock);
-    auto iter = _contexts.find(query_id);
-    if (iter != _contexts.end()) {
-        iter->second->increment_num_fragments();
-        return iter->second.get();
+    size_t i = std::hash<size_t>()(query_id.lo) & QUERY_CONTEXT_MAP_SLOT_NUM_MASK;
+    auto& mutex = _mutexes[i];
+    auto& context_map = _context_maps[i];
+    auto& sc_map = _second_chance_maps[i];
+
+    {
+        // require read lock
```

Comment:
These comments seems needn't

Reply:
for context_map,  write operations are very rare, O(num_queries).  but read operations are very often,  O(num_fragment_instances), so use read lock to reduce contest.